lib/libvfio-user.c

```diff
@@ -927,126 +901,147 @@ device_reset(vfu_ctx_t *vfu_ctx, vfu_reset_type_t reason)
 
-    assert(vfu_ctx != NULL);
-    assert(msg != NULL);
+            memcpy(msg->out.iov.iov_base, msg->in.iov.iov_base,
+                   sizeof(struct vfio_user_device_feature));
 
-    if (msg->in.iov.iov_len < sizeof(*dirty_pages) ||
-        dirty_pages->argsz < sizeof(*dirty_pages)) {
-        vfu_log(vfu_ctx, LOG_ERR, "invalid message size %zu", msg->in.iov.iov_len);
-        return ERROR_INT(EINVAL);
-    }
+            ret = migration_feature_get(vfu_ctx, feature,
+                                        msg->out.iov.iov_base +
+                                        sizeof(struct vfio_user_device_feature));
 
-    if (vfu_ctx->migration == NULL) {
-        vfu_log(vfu_ctx, LOG_ERR, "migration not configured");
-        return ERROR_INT(ENOTSUP);
-    }
+            struct vfio_user_device_feature *res = msg->out.iov.iov_base;
 
-    switch (dirty_pages->flags) {
-    case VFIO_IOMMU_DIRTY_PAGES_FLAG_START:
-        ret = dma_controller_dirty_page_logging_start(vfu_ctx->dma,
-                  migration_get_pgsize(vfu_ctx->migration));
-        break;
+            if (ret < 0) {
+                msg->out.iov.iov_len = 0;
+            } else {
+                res->argsz = sizeof(struct vfio_user_device_feature)
+                    + sizeof(struct vfio_user_device_feature_migration);
+            }
+        } else if (is_dma_feature(feature)) {
+            ssize_t bitmap_size = dma_get_request_bitmap_size(
+                req->argsz - sizeof(struct vfio_user_device_feature),
+                req->data
+            );
+
+            msg->out.iov.iov_len = sizeof(struct vfio_user_device_feature)
+                + sizeof(struct vfio_user_device_feature_dma_logging_report)
+                + bitmap_size;
+            msg->out.iov.iov_base = malloc(msg->out.iov.iov_len);
+
+            if (msg->out.iov.iov_base == NULL) {
+                return ERROR_INT(ENOMEM);
+            }
 
-    case VFIO_IOMMU_DIRTY_PAGES_FLAG_STOP:
-        dma_controller_dirty_page_logging_stop(vfu_ctx->dma);
-        ret = 0;
-        break;
+            memcpy(msg->out.iov.iov_base, msg->in.iov.iov_base,
+                   sizeof(struct vfio_user_device_feature) +
+                   sizeof(struct vfio_user_device_feature_dma_logging_report));
 
-    case VFIO_IOMMU_DIRTY_PAGES_FLAG_GET_BITMAP:
-        ret = handle_dirty_pages_get(vfu_ctx, msg);
-        break;
+            struct vfio_user_device_feature *res = msg->out.iov.iov_base;
 
-    default:
-        vfu_log(vfu_ctx, LOG_ERR, "bad flags %#x", dirty_pages->flags);
-        ret = ERROR_INT(EINVAL);
-        break;
+            res->argsz = msg->out.iov.iov_len;
+
+            ret = dma_feature_get(vfu_ctx, feature, res->data);
+
+            if (ret < 0) {
+                msg->out.iov.iov_len = 0;
+            }
+        } else {
+            return ERROR_INT(EINVAL);
+        }
+    } else if (req->flags & VFIO_DEVICE_FEATURE_SET) {
+        msg->out.iov.iov_base = malloc(msg->in.iov.iov_len);
```

Comment:
ditto

Reply:
As above, `msg->in.iov.iov_len` is safe.